src/hotspot/share/opto/type.cpp

```diff
@@ -1113,6 +1113,72 @@ void Type::dump_stats() {
 }
 #endif
 
+//------------------------------category---------------------------------------
+#ifndef PRODUCT
+Type::CATEGORY Type::category() const {
+  const TypeTuple* tuple;
+  switch (base()) {
+  case Type::Int:
+  case Type::Long:
+  case Type::Half:
+  case Type::NarrowOop:
+  case Type::NarrowKlass:
+  case Type::Array:
+  case Type::VectorA:
+  case Type::VectorS:
+  case Type::VectorD:
+  case Type::VectorX:
+  case Type::VectorY:
+  case Type::VectorZ:
+  case Type::AnyPtr:
+  case Type::RawPtr:
+  case Type::OopPtr:
+  case Type::InstPtr:
+  case Type::AryPtr:
+  case Type::MetadataPtr:
+  case Type::KlassPtr:
+  case Type::Function:
+  case Type::Return_Address:
+  case Type::FloatTop:
+  case Type::FloatCon:
+  case Type::FloatBot:
+  case Type::DoubleTop:
+  case Type::DoubleCon:
+  case Type::DoubleBot:
+    return CatData;
+  case Type::Memory:
+    return CatMemory;
+  case Type::Control:
+    return CatControl;
+  case Type::Top:
+  case Type::Abio:
+  case Type::Bottom:
+    return CatOther;
+  case Type::Bad:
+  case Type::lastype:
+    return CatUndef;
+  case Type::Tuple:
+    // Recursive case. Return CatMixed if the tuple contains types of different
+    // categories (e.g. CallStaticJavaNode's type), or the specific category if
+    // all types are of the same category (e.g. IfNode's type).
+    tuple = is_tuple();
+    if (tuple->cnt() == 0) {
+      return CatUndef;
+    } else {
+      CATEGORY first = tuple->field_at(0)->category();
+      for (uint i = 1; i < tuple->cnt(); i++) {
+        if (tuple->field_at(i)->category() != first) {
+          return CatMixed;
+        }
+      }
+      return first;
+    }
+  }
+  assert(false, "unmatched base type");
+  return CatUndef;
```

Comment:
You could add that to an explicit default case to the above switch statement to make it more clear.

Reply:
Done.